anomalib/models/cfa/torch_model.py

```diff
@@ -0,0 +1,387 @@
+
+    def compute_distance(self, target_oriented_features: Tensor) -> Tensor:
+        """Compute distance using target oriented features.
+
+        Args:
+            target_oriented_features (Tensor): Target oriented features computed
+                using the descriptor.
+
+        Returns:
+            Tensor: Distance tensor.
+        """
+        if target_oriented_features.ndim == 4:
+            target_oriented_features = rearrange(target_oriented_features, "b c h w -> b (h w) c")
+
+        features = target_oriented_features.pow(2).sum(dim=2, keepdim=True)
+        centers = self.memory_bank.pow(2).sum(dim=0, keepdim=True).to(features.device)
+        f_c = 2 * torch.matmul(target_oriented_features, (self.memory_bank.to(features.device)))
+        distance = features + centers - f_c
+        return distance
+
+    def compute_loss(self, distance: Tensor) -> Tensor:
+        """Compute the CFA loss.
+
+        Args:
+            distance (Tensor): Distance computed using target oriented features.
+
+        Returns:
+            Tensor: CFA loss.
+        """
+        num_neighbors = self.num_nearest_neighbors + self.num_hard_negative_features
+        distance = distance.topk(num_neighbors, largest=False).values
+
+        score = distance[:, :, : self.num_nearest_neighbors] - (self.radius**2).to(distance.device)
+        l_att = torch.mean(torch.max(torch.zeros_like(score), score))
+
+        score = (self.radius**2).to(distance.device) - distance[:, :, self.num_hard_negative_features :]
+        l_rep = torch.mean(torch.max(torch.zeros_like(score), score - 0.1))
+
+        loss = (l_att + l_rep) * 1000
+
+        return loss
+
+    def compute_score(self, distance: Tensor) -> Tensor:
+        """Compute score based on the distance.
+
+        Args:
+            distance (Tensor): Distance tensor computed using target oriented
+                features.
+
+        Returns:
+            Tensor: Score value.
+        """
+        distance = torch.sqrt(distance)
+
+        n_neighbors = self.num_nearest_neighbors
+        distance = distance.topk(n_neighbors, largest=False).values
+
+        distance = (F.softmin(distance, dim=-1)[:, :, 0]) * distance[:, :, 0]
+        distance = distance.unsqueeze(-1)
+
+        score = rearrange(distance, "b (h w) c -> b c h w", h=self.scale[0], w=self.scale[1])
+        return score.detach()
+
+    def compute_anomaly_map(self, score: Tensor) -> Tensor:
+        """Compute anomaly map based on the score.
+
+        Args:
+            score (Tensor): Score tensor.
+
+        Returns:
+            Tensor: Anomaly map.
+        """
+        anomaly_map = score.mean(dim=1, keepdim=True)
+        anomaly_map = F.interpolate(anomaly_map, size=self.input_size, mode="bilinear", align_corners=False)
+
+        gaussian_blur = GaussianBlur2d(sigma=4).to(score.device)
+        anomaly_map = gaussian_blur(anomaly_map)  # pylint: disable=not-callable
+        return anomaly_map
```

Comment:
Should this be wrapped into a `AnomalyMapGenerator` class to match the structure used by other Anomalib models?